pinot-tools/src/main/java/com/linkedin/pinot/tools/perf/PerfBenchmarkRunner.java

```diff
@@ -0,0 +1,168 @@
+  }
+
+  @Override
+  public boolean execute()
+      throws Exception {
+    if (_mode.equalsIgnoreCase("startAll") || _mode.equalsIgnoreCase("startAllButServer")) {
+      startAllButServer();
+    }
+    if (_mode.equalsIgnoreCase("startAll") || _mode.equalsIgnoreCase("startServerWithPreLoadedSegments")) {
+      startServerWithPreLoadedSegments();
+    }
+    return true;
+  }
+
+  public static void startAllButServer()
+      throws Exception {
+    PerfBenchmarkDriverConf perfBenchmarkDriverConf = new PerfBenchmarkDriverConf();
+    perfBenchmarkDriverConf.setStartServer(false);
+    PerfBenchmarkDriver driver = new PerfBenchmarkDriver(perfBenchmarkDriverConf);
+    driver.run();
+  }
+
+  private void startServerWithPreLoadedSegments()
+      throws Exception {
+    PerfBenchmarkDriverConf perfBenchmarkDriverConf = new PerfBenchmarkDriverConf();
+    perfBenchmarkDriverConf.setStartZookeeper(false);
+    perfBenchmarkDriverConf.setStartController(false);
+    perfBenchmarkDriverConf.setStartBroker(false);
+    perfBenchmarkDriverConf.setServerInstanceDataDir(_dataDir);
+    final PerfBenchmarkDriver driver =
+        new PerfBenchmarkDriver(perfBenchmarkDriverConf, _tempDir, _loadMode, _segmentFormatVersion, false);
+    driver.run();
+
+    if (_isBatchLoad) {
+      String[] tableNames = new File(_dataDir).list();
+      Preconditions.checkNotNull(tableNames);
+      ExecutorService executorService = Executors.newFixedThreadPool(_numThreads);
+      for (final String tableName : tableNames) {
+        executorService.submit(new Runnable() {
+          @Override
+          public void run() {
+            try {
+              loadTable(driver, _dataDir, tableName, null);
+            } catch (Exception e) {
+              LOGGER.error("Caught exception while loading table: {}", tableName, e);
+            }
+          }
+        });
+      }
+      executorService.shutdown();
+      executorService.awaitTermination(_timeoutInSeconds, TimeUnit.SECONDS);
+    } else {
+      List<String> invertedIndexColumns = null;
+      if (_invertedIndexColumns != null) {
+        invertedIndexColumns = Arrays.asList(_invertedIndexColumns.split(","));
+      }
+      for (String tableName : _tableNames.split(",")) {
+        loadTable(driver, _dataDir, tableName, invertedIndexColumns);
+      }
+    }
+  }
+
+  public static void loadTable(PerfBenchmarkDriver driver, String dataDir, String tableName,
+      List<String> invertedIndexColumns)
+      throws Exception {
+    boolean tableConfigured = false;
+    File[] segments = new File(dataDir, tableName).listFiles();
+    Preconditions.checkNotNull(segments);
+    for (File segment : segments) {
+      SegmentMetadataImpl segmentMetadata = new SegmentMetadataImpl(segment);
+      if (!tableConfigured) {
+        driver.configureTable(segmentMetadata.getTableName(), invertedIndexColumns);
+        tableConfigured = true;
+      }
+      driver.addSegment(segmentMetadata);
```

Comment:
More than multi threaded, I think this was the slowest part. Since we do one segment at a time. We should have a method to do addSegments(segmentMetadataList) or something. This will ensure that there is only one write to ZK.